### PR TITLE
Upstream/module hooks

### PR DESCRIPTION
**Pull request by hlyytine**
- This makes it possible for a VMM module to customize generated/provided DTB.
- Existing APIs are not changed.
- Existing semantics are not changed, existing code should compile/work as earlier.
- The same pattern can be used for other hooks in the future.

### CODE REVIEW

**Comment by hlyytine** [comment]
The proposal from @chrisguikema that VMM should be modular and minimizable resonates well within our team. I found tiiuae/camkes-vm#68 from @axel-h particularly hard to follow (read: not sure if I understood everything). Can you look at this and tell me what you think.

Note that if we go with this approach, we shall fix that TODO before merging, i.e., move stuff from fdt_manipulation.c to external library (or more specifically, provided externally usable header) for modules to use.

**Comment by hlyytine** [comment]
And yes, this is not an exact alternative to seL4/camkes-vm#68: it does not have mechanism how to disable init_ram module and let some other module map the guest RAM. I have some ideas, I just need to write them down.

**Comment by axel-h** [comment]
I think this makes the module interface even more complex. My preference is, that there is one function that gets a context pointer with all VMM/VM things. Then this is a clean OO approach without global variables. I will make a proposal for that.
The module dependency issue might something separate to think about still.

**Comment by hlyytine** [comment]
Yeah, after all FDT generation can itself be a module; if and when we add mechanism for modules to require some other modules to be loaded, optional FDT generation could be their own `DEFINE_MODULE`.

**Comment by hlyytine** [comment]
Actually my last comment is wrong. The DTB generation code is only after loading modules.
